### ISSUE TITLE
Remotely debugging .NET 7 apps in linux-arm environments causes segmentation fault

### ISSUE THREAD

**Issue by CamSoper** [dotnet/runtime#81921]
### Description

Unable to remotely debug .NET 7 apps in `linux-arm` (32-bit).

### Reproduction Steps

1. Create new .NET 7 console project.
2. Replace the contents of *Program.cs* with:
  ```charp
  Console.WriteLine("I'm not going to do anything!");
  while(true)
  {
      System.Threading.Thread.Sleep(1000);
  }
  ```
3. Publish the self-contained app with `dotnet publish --runtime linux-arm --self-contained` and deploy the files to the device [as described here](https://learn.microsoft.com/dotnet/iot/deployment).
4. Ensure the device has vsdbg installed: `curl -sSL https://aka.ms/getvsdbgsh | /bin/sh /dev/stdin -v latest -l ~/vsdbg`
5. From a PC, use the following VS Code launch.json configuration to remotely launch the app in the debugger:
  ```json
  {
    "name": "Remote Launch Pi",
    "type": "coreclr",
    "request": "launch",
    "program": "~/myapp/myapp",
    "args": [],
    "cwd": "~/myapp",
    "stopAtEntry": false,
    "console": "internalConsole",
    "pipeTransport": {
      "pipeCwd": "${workspaceRoot}",
      "pipeProgram": "ssh",
      "pipeArgs": [
        "pi@raspberrypi"
      ],
      "debuggerPath": "~/vsdbg/vsdbg"
    }
  ```
6. Alternatively, launch the app on the Pi and then attach to the running process from the PC with this configuration:
  ```json
  {
    "name": "Attach Pi",
    "type": "coreclr",
    "request": "attach",
    "processId": "${command:pickRemoteProcess}",
    "pipeTransport": {
    "pipeCwd": "${workspaceRoot}",
    "pipeProgram": "ssh",
    "pipeArgs": [
      "pi@raspberrypi"
    ],
    "debuggerPath": "~/vsdbg/vsdbg"
    }
  }
  ```

### Expected behavior

Application runs and the debugging session proceeds.

### Actual behavior

When attempting to launch in the debugger:

```plaintext
Starting: "ssh" pi@192.168.15.141 ~/vsdbg/vsdbg --interpreter=vscode
-------------------------------------------------------------------
You may only use the Microsoft .NET Core Debugger (vsdbg) with
Visual Studio Code, Visual Studio or Visual Studio for Mac software
to help you develop and test your applications.
-------------------------------------------------------------------
Loaded '/home/pi/Animate/System.Private.CoreLib.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
The program '[3178] Animate' has exited with code 0 (0x0).
```

If attempting to attach to a running process, the debugger attaches and then after breaking/continuing an indeterminite number of times, exits with `Segmentation Fault`.

### Regression?

This currently works in .NET 6. .NET 7 seems to be the only runtime affected.

### Known Workarounds

- Use a 64-bit OS and target `linux-arm64` (when using .NET 7)
- Use .NET 6 and target `linux-arm` on a 32-bit OS

### Configuration

Isolated to `linux-arm`. Attempted to repro on `linux-x64` and `linux-arm64`. `linux-arm` is the only runtime affected.

This behavior has been verified on 2 separate Raspberry Pi 3 devices and one Raspberry Pi 4 device, in both Raspberry Pi OS (32-bit) and Ubuntu (32-bit).

This behavior has been observed in Visual Studio Code (both launching the process and attaching to an existing process) and in Visual Studio (attaching remotely).

### Other information

This behavior has been [observed by other users](https://stackoverflow.com/questions/74994982/debug-net-apps-on-raspberry-pi-dotnet-has-exited-with-code-0).

**Comment by richlander** [comment]
Thanks for reporting this.

It is assumed (but not fully stated) from the write-up that the app works as expected w/o debugging in the picture. Is that fair?

**Comment by CamSoper** [comment]
> It is assumed (but not fully stated) from the write-up that the app works as expected w/o debugging in the picture. Is that fair?

Yep, correct. AFAIK, it's only when debugging. I only found it because I'm making a video series and I got to the part where I show how to debug and... couldn't.

**Comment by richlander** [comment]
@noahfalk

**Comment by noahfalk** [comment]
@mikem8361 - Mike I assigned this to you for the moment but load balance among the debugger crew as needed :)

**Comment by devbar** [comment]
I am facing the same issue. I can deploy self-contained app with .net 6.0 and debug, but not with dotnet 7.0.

On the other hand I am not able to debug with framework-dependend deployment. It doesn't matter if I use .net 6.0 or .net 7.0 here.

1. Running my build with `dotnet publish --runtime linux-arm --self-contained`

csproj:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

</Project>
```
2. Copy to Raspi with:
`scp -r .\bin\Debug\net7.0\linux-arm\publish\* pi@raspi:/home/pi/BluettiBridge/`

3. Launch debug with:
```
{
          "name": ".NET Remote Launch - Self-contained",
          "type": "coreclr",
          "request": "launch",
          "program": "~/BluettiBridge/BluettiBridge",            
          "args": [],
          "cwd": "~/BluettiBridge",
          "stopAtEntry": false,
          "console": "internalConsole",
          "pipeTransport": {
              "pipeCwd": "${workspaceRoot}",
              "pipeProgram": "ssh",
              "pipeArgs": [
                  "pi@hebe.leviathan.lan"
              ],
              "debuggerPath": "~/vsdbg/vsdbg"
          }
      }
```

Output is:

```
Wird gestartet: "ssh" pi@raspi "~/vsdbg/vsdbg --interpreter=vscode"
-------------------------------------------------------------------
You may only use the Microsoft .NET Core Debugger (vsdbg) with
Visual Studio Code, Visual Studio or Visual Studio for Mac software
to help you develop and test your applications.
-------------------------------------------------------------------
Loaded '/home/pi/BluettiBridge/System.Private.CoreLib.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
The program '[2809] BluettiBridge' has exited with code 0 (0x0).
```

----

Dev Machine:
- Win10 22H2
- VsCode 1.77.3

Target Machine:
```
processor       : 0
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 108.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd08
CPU revision    : 3

processor       : 1
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 108.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd08
CPU revision    : 3

processor       : 2
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 108.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd08
CPU revision    : 3

processor       : 3
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 108.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd08
CPU revision    : 3

Hardware        : BCM2711
Revision        : c03112
Serial          : 100000004a82e1f0
Model           : Raspberry Pi 4 Model B Rev 1.2
```

```
No LSB modules are available.
Distributor ID: Raspbian
Description:    Raspbian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye
```

```
dotnet --list-sdks: 7.0.305 [/home/pi/dotnet/sdk]

dotnet --list-runtimes: 
Microsoft.AspNetCore.App 7.0.8 [/home/pi/dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.NETCore.App 7.0.8 [/home/pi/dotnet/shared/Microsoft.NETCore.App]

dotnet-version: 
7.0.305
```

**Comment by hoyosjs** [comment]
@devbar What was the expectation? Was there a message about segfaulting? Was there a break isntruction?

**Comment by CamSoper** [comment]
Sorry for the slow reply @tommcdon. I'll dig out my test hardware tomorrow and see if it still repros

**Comment by CamSoper** [comment]
@tommcdon I finally got around to digging out my hardware. `linux-arm` builds immediately exit when the debugger is attached when the assembly is built using .NET 7 and .NET 8. Using .NET 6, the debugger successfully attaches and the program runs as expected.

I saw you changed the milestone to `9.0.0`, wasn't @richlander talking about dropping `linux-arm` support in .NET 9?

**Comment by Pearo** [comment]
I am having the same issue. Started a couple of releases ago, but I cant pinpoint which one. I cant roll back to .net 6 as I have too much .net 7 specific code.

I just get a segfault in the console when the debugger connects.

Microsoft Visual Studio Professional 2022 (64-bit) - Current
Version 17.7.2
OS Name	Microsoft Windows 11 Pro
Version	10.0.22000 Build 22000

If you need anymore info let me know.

**Comment by tommcdon** [comment]
Reactivating for a backport to .NET 8

**Comment by tommcdon** [comment]
Merged into release/8.0
